File: interview/coding-interview/bitonic_array_maximum.cpp
#include<iostream>

using namespace std;

int findMax(int *arr, int size) {
  //int len = *(&arr + 1) - arr;

  int start = 0, end = size - 1;
  
  while (start < end) {
    int mid = start + (end - start) / 2;
    if (arr[mid] > arr[mid + 1]) 
      end = mid;
    else 
      start = mid + 1;
  }
  return arr[start];
}

int main() {
  int arr[] = {1, 3, 8, 12, 4, 2};
  //int size = sizeof(arr)/sizeof(arr[0]);
  int size = *(&arr + 1) - arr;
  cout << findMax(arr, size) << endl;

}

File: interview/python/bitonic_array_maximum.py

# input is a sorted (array, list, or matrix), use binary search or two pointer
# strategy
def findMax(arr):
    start = 0
    end = len(arr) - 1
    while start < end:
        mid = (start + end) // 2
        if (arr[mid] < arr[mid + 1]):
            start = mid + 1;
        else:
            end = mid;
    return end;

def main():
    print(findMax([1, 3, 8, 12, 4, 2]))

def test1():
    assert findMax([1, 3, 8, 12, 4, 2]) == 3 
    assert findMax([1, 2, 3, 1]) == 2 
